### ISSUE TITLE
Password length requirements no longer meet NZISM

### ISSUE THREAD

**Issue by chillu** [silverstripe/cwp-core#52]
We require 8 chars min length, but NZISM requires 10 since at least v2.4 (Nov 2015) - see [archive](https://www.nzism.gcsb.govt.nz/home/previous-versions/).

Context: https://github.com/silverstripe/silverstripe-framework/issues/8522. Tracking this separately for CWP, because we might need to handle this faster or in a different way. In open source, it's a matter of opinion. In CWP, it's a compliance issue. Note that agencies can choose to increase those requirements on their own instances, but is about setting compliant defaults.

https://www.nzism.gcsb.govt.nz/ism-document/#1858

> Agencies SHOULD implement a password policy enforcing either:
> a minimum password length o
f 16 characters with no complexity requirement; or
> a minimum password length of ten characters, consisting of at least three of the following character sets:
> lowercase characters (a-z);
> uppercase characters (A-Z);
> digits (0-9); and
> punctuation and special characters.

This also aligns with OWASP guidance: https://www.owasp.org/index.php/Authentication_Cheat_Sheet#Password_Length

https://github.com/silverstripe/cwp-core/commit/217ddec2e26ea839e1549425d7e07eb5517a5df7#diff-8a7315557cd9f672e36f7e8f0ce2c29e removed the password requirements from CWP, because they're now part of recipe-core: https://github.com/silverstripe/recipe-core/blob/4/app/_config.php#L8

## Pull requests

- [x] #53 
- [x] https://github.com/silverstripe/recipe-core/pull/34
- [x] https://github.com/silverstripe/silverstripe-framework/pull/8587
- [x] https://github.com/silverstripe/cwp-core/pull/59
- [x] https://github.com/silverstripe/silverstripe-framework/pull/8600
- [x] https://github.com/silverstripe/recipe-core/pull/35
- [x] https://github.com/silverstripe/silverstripe-framework/pull/8604

**Comment by chillu** [comment]
We might need to make PasswordValidator configurable via YAML instead of using `Member::set_password_validator($validator)` to allow for this. The class already uses `config()`, so it's just a matter of NOT using a custom instance and falling back to YAML.

**Comment by sminnee** [comment]
We’d also want to think about the UX for if someone logs in with a no-longer-compliant password. Do we force a reset?

**Comment by chillu** [comment]
@rupachup Why is this in 4.3.0? Reassigning to 4.4.0

**Comment by ScopeyNZ** [comment]
It's not really core team - We should try to get this in for CWP 2.2. We can just remove the milestone.

**Comment by NightJar** [comment]
PR #53

**Comment by NightJar** [comment]
I've raised your comment @sminnee as a separate issue - #54

**Comment by robbieaverill** [comment]
Not fixed in CWP 2.2.0-rc1

**Comment by robbieaverill** [comment]
The PR was merged into master, which is probably why this is "not fixed". I'm just setting up a local environment to verify that before I cherry pick it into 2.2

**Comment by NightJar** [comment]
oh, whoops! That's my bad.

**Comment by robbieaverill** [comment]
https://github.com/silverstripe/recipe-core/blob/4.3/app/_config.php this file is coming through because we don't have an equivalent file to override it with. The config in that file is taking precedence over the YAML in #53

**Comment by robbieaverill** [comment]
Sorry, that recipe patch hasn't fixed it yet. The silverstripe/recipe-core file is still being copied in first and taking precedence.

**Comment by robbieaverill** [comment]
The problem is that this file is not extensible: https://github.com/silverstripe/recipe-core/blob/4.3/app/_config.php

It gets loaded first by composer, and the file gets copied into app/_config.php first, then all subsequent recipes will ignore their own _config.php files because it already exists. The logic in this file is executed after the config manifests are built, and after all module _config.php files, so we can't override it at the moment.

I thought we could move the password requirements from a PHP file into YAML config, but since it'll exist in user project code it'd also get loaded after modules, so the CWP config would need to say `After: '#myprojectpasswords'` or something, and since that file is in user code it could have its name changed easily.

I'm still thinking about how to get around this.

**Comment by robbieaverill** [comment]
Ok new pull requests are up, ready for discussion

**Comment by robbieaverill** [comment]
This probably need a docs update, will do that shortly

**Comment by robbieaverill** [comment]
Docs PR added https://github.com/silverstripe/silverstripe-framework/pull/8600

**Comment by robbieaverill** [comment]
I found a bug, new PR at https://github.com/silverstripe/recipe-core/pull/35

**Comment by NightJar** [comment]
🤔 I would have thought that [framework using injector if one is not set](https://github.com/silverstripe/silverstripe-framework/blob/1944a5a99b1c02c497676f6872882a2cf2250bea/src/Security/Member.php#L434-L440) should cover that...

**Comment by robbieaverill** [comment]
@NightJar yeah, but Injector doesn't `has()` one until one has been instantiated via injector. I've added that back in in silverstripe/recipe-core#35

**Comment by NightJar** [comment]
True, but [here](https://github.com/creative-commoners/cwp-core/blob/e9baf31996e5a4bc1d6cde6c6bd06b988eb864cf/_config/security.yml#L13) it sets config so it [does `has`](https://github.com/silverstripe/silverstripe-framework/blob/1944a5a99b1c02c497676f6872882a2cf2250bea/src/Core/Injector/Injector.php#L1065-L1072) one 🤔 

To be clear I'm not particularly caring about reintroducing the setting method in https://github.com/silverstripe/recipe-core/pull/35 - I'm just puzzled as to why it's necessary - it seems like it shouldn't be (if nothing else replaces the defined service linked above)... I wonder if there's anything deeper at play.

**Comment by robbieaverill** [comment]
@NightJar injector has() won’t return true with config only, it needs to have an instance created already for it to return true. Config only is not enough

**Comment by robbieaverill** [comment]
Here's a unit test to show what I mean:

```php
public function testInjectorHas()
{
    $this->assertFalse(Injector::inst()->has(PasswordValidator::class));

    Member::set_password_validator(PasswordValidator::singleton());
    $this->assertTrue(Injector::inst()->has(PasswordValidator::class));
}
```

**Comment by NightJar** [comment]
> @NightJar injector has() won’t return true with config only, it needs to have an instance created already for it to return true. Config only is not enough

If this is true then there's a bug in framework.

`has` calls `getServiceName`
`getServiceName` calls `getServiceSpec`
`getServiceSpec` will return true if there is a registered spec, not an instance.

If there is _not_ an already loaded spec, then `getServiceSpec` [performs a lazy load](https://github.com/silverstripe/silverstripe-framework/blob/4.3.0-rc1/src/Core/Injector/Injector.php#L1071) which involves:
`configLocator->locateConfigFor` fetches the spec
`load` registers the spec - and replaces the instance in memory if and only if there is already an instance in memory.

So in both cases `has` should return true if a _spec_ (i.e. config) is loaded, and not be concerned with instances.

So for the configuration in `passwords.yml` to not be loaded at the time the `assertFalse` is run, something must surely be wrong elsewhere.
Say like... [not being included in `4.3.0-rc1`](https://github.com/silverstripe/silverstripe-framework/tree/4.3.0-rc1/_config)

**Comment by robbieaverill** [comment]
All PRs merged again =)
